### PR TITLE
Add SystemdMeter

### PR DESCRIPTION
**Pull request by cgzones**
systemd_meter

### CODE REVIEW

**Comment by fasterit** [comment]
Why not parse `systemctl show` which would not add a libsystemd dependency and works runtime?

**Comment by fasterit** [comment]
`systemctl show --property=NInstalledJobs --property=NFailedJobs --property=NNames --property=NJobs` for the ones you currently get via DBus

File: SystemdMeter.c [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+   free(systemState);
+   systemState = NULL;
+   nFailedUnits = nInstalledJobs = nNames = nJobs = (uint32_t)-1;
+
+   /* Connect to the system bus */
+   r = sd_bus_open_system(&bus);
+   if (r < 0) {
+      //fprintf(stderr, "Failed to connect to system bus: %s\n", strerror(-r));
+      goto finish;
+   }
```
Comment (BenBE) [inline SystemdMeter.c:36]: What about connecting once at startup?

File: SystemdMeter.c [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+   r = sd_bus_get_property_string(bus,
+                                  "org.freedesktop.systemd1",           /* service to contact */
+                                  "/org/freedesktop/systemd1",          /* object path */
+                                  "org.freedesktop.systemd1.Manager",   /* interface name */
+                                  "SystemState",                        /* property name */
+                                  &error,                               /* object to return error in */
+                                  &systemState);
+   if (r < 0) {
+      //fprintf(stderr, "Failed to issue method call: %s\n", error.message);
```
Comment (BenBE) [inline SystemdMeter.c:46]: Should this debug code really be kept? (Same for the other locations)

File: SystemdMeter.c [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+#include <systemd/sd-bus.h>
+
+static const int SystemdMeter_attributes[] = {
+   SYSTEMD_FAILED,
+   SYSTEMD_NOTICE,
+   SYSTEMD_OK
+};
+
+static char *systemState = NULL;
+static uint32_t nFailedUnits = 0, nInstalledJobs = 0, nNames = 0, nJobs = 0;
```
Comment (BenBE) [inline SystemdMeter.c:14]: Shouldn't this be part of the meter local data (`values` array)?

File: DiskIOMeter.h [hunk 4]
```diff
@@ -2,7 +2,7 @@
 #define HEADER_DiskIOMeter
 /*
  h top - DiskIOMeter*.h
-(C) 2020 Christian Göttsche
+(C) 2020 Christian Goettsche
```
Comment (BenBE) [inline DiskIOMeter.h:5]: Unrelated change?

File: DiskIOMeter.c [hunk 5]
```diff
@@ -1,6 +1,6 @@
 /*
 htop - DiskIOMeter.c
-(C) 2020 Christian Göttsche
+(C) 2020 Christian Goettsche
```
Comment (BenBE) [inline DiskIOMeter.c:3]: Unrelated change

**Comment by cgzones** [comment]
Re-implemented via executing `systemctl`

**Comment by BenBE** [comment]
Personally I don't like having to rely on parsing the output of external programs when a proper API for it exists. Even more so if this external API probably is much faster to access than repeatedly starting a new process.

**Comment by fasterit** [comment]
Please add (c) headers to the new files.
Otherwise LGTM. Please feel free to merge.

**Comment by fasterit** [comment]
> Personally I don't like having to rely on parsing the output of external programs when a proper API for it exists. Even more so if this external API probably is much faster to access than repeatedly starting a new process.

But this way we do not need to hard-depend on libsystemd which would mean either of:

- htop variants
- htop without systemd support
- htop pulling in systemd

for distros that support multiple init systems.

And tbh., we parse many more strings than these...

File: SystemdMeter.c [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+#include "StringUtils.h"
+
+static const int SystemdMeter_attributes[] = {
+   SYSTEMD_FAILED,
+   SYSTEMD_NOTICE,
+   SYSTEMD_OK
+};
+
+static char *systemState = NULL;
+static unsigned int nFailedUnits = 0, nInstalledJobs = 0, nNames = 0, nJobs = 0;
```
Comment (BenBE) [inline SystemdMeter.c:20]: Shouldn't this be `(unsigned int)-1` as used in other places for undefined values?

**Comment by BenBE** [comment]
> > Personally I don't like having to rely on parsing the output of external programs when a proper API for it exists. Even more so if this external API probably is much faster to access than repeatedly starting a new process.
> 
> But this way we do not need to hard-depend on libsystemd which would mean either of:
> 
> * htop variants
> * htop without systemd support
> * htop pulling in systemd
> 
> for distros that support multiple init systems.

We still could `dlopen`/`dlsym` it without causing the link time dependency. Wouldn't even be much overhead code-wise to do.

> And tbh., we parse many more strings than these...

Good point. :grinning:

**Comment by fasterit** [comment]
> We still could `dlopen`/`dlsym` it without causing the link time dependency. Wouldn't even be much overhead code-wise to do.

That would be fine for me as well.
I measured the overhead of the execl call and it is negligible on an i7 CPU. If somebody want to measure on their ARMv5, feel free :-).

**Comment by pthfdr-42** [comment]
Please not.
Doing this would likely to let distros to make systemd as a dependency of htop.

**Comment by fasterit** [comment]
> Please not.
> Doing this would likely to let distros to make systemd as a dependency of htop.

That's not true. The last two iterations of this PR soft depend on systemd only, i.e. you can use the meter if you have systemd running anyways but htop does not depend in an installed libsystemd otherwise. That was a deliberate choice especially to not force systemd on people that do not want to run it.

**Comment by fasterit** [comment]
1. The systemd meter only makes sense on Linux.
2. Is there a use case where systemd is active but the dlopen call fails and you need to do the parsing of `systemctl` output?
     I suggested the latter only to prevent the hard dependency on libsystemd. That is already prevented by dlopen use now.

**Comment by cgzones** [comment]
On minimal systems without D-Bus the library method does not work.
There systemctl succeeds by using */run/systemd/private*, which does not have a stability promise: https://systemd.io/PORTABILITY_AND_STABILITY/

**Comment by fasterit** [comment]
> On minimal systems without D-Bus the library method does not work.
> There systemctl succeeds by using _/run/systemd/private_, which does not have a stability promise: https://systemd.io/PORTABILITY_AND_STABILITY/

thanks, that resolves the question 2. from above

File: linux/SystemdMeter.c [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+*/
+
+#include "SystemdMeter.h"
+
+#include <dlfcn.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
```
Comment (BenBE) [inline linux/SystemdMeter.c:15]: After `fcntl.h`

File: linux/SystemdMeter.c [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+      if (dlerror() != NULL)
+         goto dlfailure;
+
+      *(void **)(&sym_sd_bus_get_property_trivial) = dlsym(dlopenHandle, "sd_bus_get_property_trivial");
+      if (dlerror() != NULL)
+         goto dlfailure;
+
+      *(void **)(&sym_sd_bus_unref) = dlsym(dlopenHandle, "sd_bus_unref");
+      if (dlerror() != NULL)
+         goto dlfailure;
```
Comment (BenBE) [inline linux/SystemdMeter.c:76]: `dlsym` may return `NULL` without indicating an error, thus the resolved symbols should be checked to be non-NULL.

File: configure.ac [hunk 9]
```diff
@@ -83,6 +83,8 @@ AC_FUNC_CLOSEDIR_VOID
 AC_FUNC_STAT
 AC_CHECK_FUNCS([memmove strncasecmp strstr strdup])
 
+AC_SEARCH_LIBS([dlopen], [dl dld])
```
Comment (BenBE) [inline configure.ac:86]: Probably should also check `dlsym`, `dlclose` and `dlerror`.

**Review by BenBE — CHANGES_REQUESTED** [review]

File: linux/SystemdMeter.c (hunk 8 above)
Comment (cgzones) [reply linux/SystemdMeter.c:76]: Isn't it the other way round? (it might return NULL on success)

https://manpages.debian.org/unstable/manpages-dev/dlsym.3.en.html
> There are several scenarios when the address of a global symbol is NULL. For example, a symbol can be placed at zero address by the linker, via a linker script or with --defsym command-line option. Undefined weak symbols also have NULL value. Finally, the symbol value may be the result of a GNU indirect function (IFUNC) resolver function that returns NULL as the resolved value. In the latter case, dlsym() also returns NULL without error. However, in the former two cases, the behavior of GNU dynamic linker is inconsistent: relocation processing succeeds and the symbol can be observed to have NULL value, but dlsym() fails and dlerror() indicates a lookup error.

Comment (BenBE) [reply linux/SystemdMeter.c:76]: Exactly. Bad phrasing on my side. ;-)

Comment (cgzones) [reply linux/SystemdMeter.c]: ahh, i missed your point at first. resolved

File: configure.ac (hunk 9 above)
Comment (cgzones) [reply configure.ac:86]: It's mainly not about checking if `dlopen()` exists, but if we need to link with `-ldl` or `-ldld` to make the function available.

File: linux/SystemdMeter.c [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+      if (!sym_sd_bus_get_property_string || dlerror() != NULL)
+         goto dlfailure;
+
+      *(void **)(&sym_sd_bus_get_property_trivial) = dlsym(dlopenHandle, "sd_bus_get_property_trivial");
+      if (!sym_sd_bus_get_property_trivial || dlerror() != NULL)
+         goto dlfailure;
+
+      *(void **)(&sym_sd_bus_unref) = dlsym(dlopenHandle, "sd_bus_unref");
+      if (!sym_sd_bus_unref || dlerror() != NULL)
+         goto dlfailure;
```
Comment (BenBE) [inline linux/SystemdMeter.c:76]:
```suggestion
#define resolve(symbolname) do {\
      *(void **)(&sym##symbolname) = dlsym(dlopenHandle, #symbolname); \
      if (!sym##symbolname || dlerror() != NULL) \
         goto dlfailure;

      resolve(sd_bus_open_system);
      resolve(sd_bus_get_property_string);
      resolve(sd_bus_get_property_trivial);
      resolve(sd_bus_unref);

#undef resolve
```

File: linux/SystemdMeter.c [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+                                      "SystemState",                        /* property name */
+                                      NULL,                                 /* object to return error in */
+                                      &systemState);
+   if (r < 0)
+      goto busfailure;
+
+   r = sym_sd_bus_get_property_trivial(bus,
+                                       "org.freedesktop.systemd1",           /* service to contact */
+                                       "/org/freedesktop/systemd1",          /* object path */
+                                       "org.freedesktop.systemd1.Manager",   /* interface name */
```
Comment (BenBE) [inline linux/SystemdMeter.c:101]: What about having constants for those 3 strings?

File: linux/SystemdMeter.c (hunk 10 above)
Comment (cgzones) [reply linux/SystemdMeter.c]: adopted

File: linux/SystemdMeter.c (hunk 11 above)
Comment (cgzones) [reply linux/SystemdMeter.c]: adopted

File: linux/SystemdMeter.c [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+
+   if (bus && dlopenHandle) {
+      sym_sd_bus_unref(bus);
+   }
+   bus = NULL;
+
+   if (dlopenHandle) {
+      dlclose(dlopenHandle);
+      dlopenHandle = NULL;
+   }
```
[the remark this answers is not in the document]
Comment (cgzones) [reply linux/SystemdMeter.c:50]: My intention was to set `bus` to NULL in every case.

[the remark this answers is not in the document]
Comment (BenBE) [reply linux/SystemdMeter.c:50]: If we ensure `bus` can only ever be non-NULL if all function pointers resolved successfully, I'm fine with dropping the `dlopenhandle` check in the first of your condition blocks.
